peer/upstream/upstream.go

```diff
@@ -0,0 +1,350 @@
+	return u.client
+}
+
+// GetSavedHeight - fetch height from last polled value
+func (u *Upstream) GetSavedHeight() uint64 {
+	return u.blockHeight
+}
+
+// GetBlockDigest - fetch block digest from a specific block number
+func (u *Upstream) GetBlockDigest(blockNumber uint64) (blockdigest.Digest, error) {
+	remoteAddr, _ := u.RemoteAddr()
+	u.log.Debugf("remote %s, get block digest %d", remoteAddr, blockNumber)
+	parameter := make([]byte, 8)
+	binary.BigEndian.PutUint64(parameter, blockNumber)
+
+	// critical section - lock out the runner process
+	u.Lock()
+	var data [][]byte
+	err := u.client.Send("H", parameter)
+	if nil == err {
+		data, err = u.client.Receive(0)
+	}
+	u.Unlock()
+
+	if nil != err {
+		return blockdigest.Digest{}, err
+	}
+
+	if 2 != len(data) {
+		return blockdigest.Digest{}, fault.ErrInvalidPeerResponse
+	}
+
+	switch string(data[0]) {
+	case "E":
+		return blockdigest.Digest{}, fault.InvalidError(string(data[1]))
+	case "H":
+		d := blockdigest.Digest{}
+		if blockdigest.Length == len(data[1]) {
+			err := blockdigest.DigestFromBytes(&d, data[1])
+			return d, err
+		}
+	default:
+	}
+	return blockdigest.Digest{}, fault.ErrInvalidPeerResponse
+}
+
+// GetBlockData - fetch block data from a specific block number
+func (u *Upstream) GetBlockData(blockNumber uint64) ([]byte, error) {
+	parameter := make([]byte, 8)
+	binary.BigEndian.PutUint64(parameter, blockNumber)
+
+	// critical section - lock out the runner process
+	u.Lock()
+	var data [][]byte
+	err := u.client.Send("B", parameter)
+	if nil == err {
+		data, err = u.client.Receive(0)
+	}
+	u.Unlock()
+
+	if nil != err {
+		return nil, err
+	}
+
+	if 2 != len(data) {
+		return nil, fault.ErrInvalidPeerResponse
+	}
+
+	switch string(data[0]) {
+	case "E":
+		return nil, fault.InvalidError(string(data[1]))
+	case "B":
+		return data[1], nil
+	default:
+	}
+	return nil, fault.ErrInvalidPeerResponse
+}
+
+// must have lock held before calling
+func (u *Upstream) GetRemoteHeight() (uint64, error) {
+	u.log.Infof("getHeight: client: %s", u.client)
+
+	err := u.client.Send("N")
+	if nil != err {
+		u.log.Errorf("getHeight: %s send error: %s", u.client, err)
+		return 0, err
+	}
+
+	data, err := u.client.Receive(0)
+	if nil != err {
+		u.log.Errorf("push: %s receive error: %s", u.client, err)
+		return 0, err
+	}
+	if 2 != len(data) {
+		return 0, fmt.Errorf("getHeight received: %d  expected: 2", len(data))
+	}
+
+	switch string(data[0]) {
+	case "E":
+		return 0, fmt.Errorf("rpc error response: %q", data[1])
+	case "N":
+		if 8 != len(data[1]) {
+			return 0, fmt.Errorf("highestBlock: rpc invalid response: %q", data[1])
+		}
+		height := binary.BigEndian.Uint64(data[1])
+		u.updateLastResponseTime()
+		u.log.Infof("height: %d", height)
+		return height, nil
+	default:
+		return 0, fmt.Errorf("rpc unexpected response: %q", data[0])
+	}
+}
+
+func (u *Upstream) updateLastResponseTime() {
+	u.lastResponseTime = time.Now()
+}
+
+func (u *Upstream) saveRemoteHeight(height uint64) {
```

Comment:
I think private setter is unnecessary unless you do complex logic inside that

Reply:
in this case not much difference

Reply:
So i think it's unnecessary. Just access the field directly.

Reply:
could be, i am still having some minor changes, it at last this method is just setter than i shall remove it